text/0249-v2-experiments.md

```diff
@@ -373,6 +367,363 @@ Disadvantages:
+npm ERR! ERESOLVE unable to resolve dependency tree
+npm ERR!
+npm ERR! While resolving: my-cdk-application@1.0.0
+npm ERR! Found: aws-cdk-lib@1.61.0-experimental
+npm ERR! node_modules/aws-cdk-lib
+npm ERR!   aws-cdk-lib"@1.61.0-experimental" from the root project
+npm ERR!
+npm ERR! Could not resolve dependency:
+npm ERR! peer aws-cdk-li@"^1.60.0" from my-construct-lib@1.6.0
+npm ERR! node_modules/my-construct-lib
+npm ERR!   my-construct-lib@"1.6.0" from the root project
+npm ERR!
+npm ERR! Fix the upstream dependency conflict, or retry
+npm ERR! this command with --force, or --legacy-peer-deps
+npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
+npm ERR!
+```
+
+Users can work around this by executing `npm install --force` which means that
+npm will not check for version compatibility at all, and therefore not an
+acceptable solution.
+
+##### Separate major versions
+
+This was rejected due to similar reasons as listed in the disadvantages of
+[Option 4: separate V3 that's all unstable](#option-4-separate-v3-thats-all-unstable).
+
+### Option 6: API Previews
+
+AWS CDK v2 release notes:
+
+> Starting with version 2.0.0 of the AWS CDK, all modules and members will
+> become stable. This means that from this release, we are committed to never
+> introduce breaking changes in a non-major bump.
+>
+> One of the most common feedback we hear from customers is that they love how
+> fast new features are added to the AWS CDK, we love it to. In v1, the
+> mechanism that allowed us to add new features quickly was marking them as
+> "experimental". Experimental features were not subject to semantic versioning,
+> and we allowed breaking changes to be introduced in these APIs. This is the
+> other most common feedback we hear from customers - breaking changes are not
+> ok.
+>
+> #### Introducing API Previews
+>
+> To make sure we can keep adding features fast, while keeping our commitment to
+> not release breaking changes, we are introducing a new model - API Previews.
+> APIs that we want to get in front of developers early, and are not yet
+> finalized, will be added to the AWS CDK with a specific suffix: `PreX`. APIs
+> with the preview suffix will never be removed, instead they will be deprecated
+> and replaced by either the stable version (without the suffix), or by a newer
+> preview version. For example, assume we add the method
+> `grantAwesomePowerPre1`:
+>
+> ```ts
+> /**
+>  * This methods grants awesome powers
+>  */
+> grantAwesomePowerPre1();
+> ```
+>
+> Times goes by, we get feedback that this method will actually be much better
+> if it accept a `Principal`. Since adding a required property is a breaking
+> change, we will add `grantAwesomePowerPre2()` and deprecate
+> `grantAwesomePowerPre1`:
+>
+> ```ts
+> /**
+> * This methods grants awesome powers to the given principal
+> *
+> * @param grantee The principal to grant powers to
+> */
+> grantAwesomePowerPre2(grantee: iam.IGrantable)
```

Comment:
I would like to see how a (new) experimental construct in a stable library would look like in each of our major languages. Could we add something like this as an appendix item?

Let's say for instance, that a new experimental `Function` construct is being introduced into a stable construct library `aws-lambda` (hypothetically). Would the API look something like this?

```ts
const handler = new FunctionPre1({
  handlerPre1: 'index.handler',
  runtimePre3: RuntimePre2.NODEJS_12_X_PRE2,
  code: CodePre1.fromAssetPre5(...),
  ...
});
```

Similarly, how would this be in Java?

Reply:
Im going to add all this in the specification doc, it will be easier to review it outside of this RFC.

Reply:
Sounds good.